smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpProtocolClientGenerator.kt

```diff
@@ -282,8 +280,33 @@ abstract class HttpProtocolClientGenerator(
      * Render any additional methods to support client operation
      */
     protected open fun renderAdditionalMethods(writer: KotlinWriter) { }
-}
 
-// TODO https://github.com/awslabs/aws-sdk-kotlin/issues/557
-private fun OperationShape.checksumRequired(): Boolean =
-    hasTrait<HttpChecksumRequiredTrait>() || getTrait<HttpChecksumTrait>()?.isRequestChecksumRequired == true
+    /**
+     * Render optionally installing Md5ChecksumMiddleware.
+     * The Md5 middleware will only be installed if the operation requires a checksum and the user has not opted-in to flexible checksums.
+     */
+    private fun OperationShape.renderIsMd5ChecksumRequired(writer: KotlinWriter) {
+        val httpChecksumTrait = getTrait<HttpChecksumTrait>()
+
+        // the checksum requirement can be modeled in either HttpChecksumTrait's `requestChecksumRequired` or the HttpChecksumRequired trait
+        if (!hasTrait<HttpChecksumRequiredTrait>() && httpChecksumTrait == null) {
+            return
+        }
+
+        val checksumAlgorithm = ctx.model.getShape(input.get()).getOrNull()
+            ?.members()
+            ?.firstOrNull { it.memberName == httpChecksumTrait?.requestAlgorithmMember?.getOrNull() }
+
+        if (hasTrait<HttpChecksumRequiredTrait>() || httpChecksumTrait?.isRequestChecksumRequired == true) {
+            writer.addImport(RuntimeTypes.Http.Middlware.Md5ChecksumMiddleware)
+
+            checksumAlgorithm?.let {
+                writer.withBlock("if (input.#L == null) {", "}", checksumAlgorithm.defaultName()) {
+                    writer.write("op.install(#T())", RuntimeTypes.Http.Middlware.Md5ChecksumMiddleware)
+                }
```

Comment:
**Question**: This installs the MD5 checksum middleware if the input has no checksum algorithm _at invocation time_? Isn't it possible for the checksum algorithm to be set later by an interceptor? Would we still want the MD5 checksum middleware to run in that case?

Reply:
Good point. Maybe we should discuss this